_source/_posts/2022-02-28-spring-native-jhipster.adoc

```diff
@@ -0,0 +1,451 @@
+npm i react-images
+----
+
+Then, you'll need to add type hints for Drew Noake's EXIF processing library in your main application class:
+
+[source,java]
+----
+@org.springframework.nativex.hint.TypeHint(
+types = {
+    ...
+    com.drew.metadata.exif.ExifIFD0Directory.class,
+    com.drew.metadata.exif.ExifSubIFDDirectory.class,
+    com.drew.metadata.exif.makernotes.AppleMakernoteDirectory.class,
+    com.drew.metadata.exif.GpsDirectory.class,
+})
+@org.springframework.nativex.hint.NativeHint(options = "-H:+AddAllCharsets")
+----
+
+The `@NativeHint` solves the following exception that happens when you upload a photo:
+
+----
+Caused by: java.nio.charset.UnsupportedCharsetException: Cp1252
+at java.nio.charset.Charset.forName(Charset.java:528) ~[native-executable:na]
+at com.drew.lang.Charsets.<clinit>(Charsets.java:40) ~[native-executable:2.16.0]
+----
+
+Once you've made all the changes (or cloned the `spring-native` branch), you can build your hip native binary.
+
+=== Build a native JHipster app
+
+You will need a JDK with GraalVM and its `native-image` compiler. Using SDKMAN, run the following command and set it as the default:
+
+[source,shell]
+----
+sdk install java 21.3.0.r17-grl
+----
+
+Add the native extension to the JDK:
+
+[source,shell]
+----
+gu install native-image
+----
+
+Then, use Maven to build the project. Skip tests since there's no support for Mockito at this time.
+
+[source,shell]
+----
+./mvnw package -Pnative,prod -DskipTests
+----
+
+This process will take several minutes to complete.
+
+=== Configure your OpenID Connect identity provider
+
+When you generate a JHipster app with OAuth 2.0 / OIDC for authentication, it defaults to using Keycloak. It creates a `src/main/docker/keycloak.yml` file for Docker Compose, as well as a `src/main/docker/realm-config` directory with files to auto-create users and OIDC clients.
+
+If you want to use Keycloak for your running app, start it with the following command:
+
+[source,shell]
+----
+docker-compose -f src/main/docker/keycloak.yml up -d
+----
+
+If you'd rather use Okta or Auth0, that's possible too!
+
+==== Use Okta as your identity provider
+
+{% include setup/cli.md type="jhipster" %}
+
+That's it!
+
+You'll need to source the `.okta.env` file the Okta CLI creates to override the default Spring Security settings.
+
+[source,shell]
+----
+source .okta.env
+----
+
+NOTE: If you're on Windows, you can modify this file to use `set` instead of `export` and rename it to `okta.bat`. Then, run it with `okta.bat` from the command line.
+
+CAUTION: Modify your existing `.gitignore` file to have `*.env` so you don't accidentally check in your secrets!
+
+_Skip to <<Run your native JHipster app>> if you've configured your app for Okta and just want to see it running._
+
+==== Use Auth0 as your identity provider
+
+To switch from Keycloak to Auth0, you only need to override the Spring Security OAuth properties. You don't even need to write any code!
```

Comment:
```suggestion
To switch from Keycloak to Auth0, override the Spring Security OAuth properties. You don't even need to write any code!
```
streamlining

Reply:
At one point, I thought about publishing this post on the Auth0 blog first, then syndicating it on ours. However, I didn't want to go through their slow editing process. Maybe we should try the reverse where we run on our blog and syndicate on theirs?